### ISSUE TITLE
Step sequencer double-click

### ISSUE THREAD

**Issue by Spekular** [LMMS/lmms#1682]
**Edit @tresf Jan 30th updated description to be more clear**

The default double-click on the step sequencer still selects a step when using the shortcut to open a piano roll.

![image](https://cloud.githubusercontent.com/assets/6345473/5981207/0be2ec4c-a886-11e4-8ef0-74956b96c64c.png)

Instead the step sequencer should only select a step when single-click is performed.  On double-click it should ignore the step sequencer and open straight to a piano roll.

---

Double clicking to open a BB-pattern in piano roll is rather clumsy. It's easy to accidentally do by adding a hit/note and then trying to remove it, and doing so on purpose often creates a hit/note. Right clicking is still slower so perhaps this ought to be reworked/replaced rather than just removed. Maybe something for @Sti2nd 's issue #1488?

**Comment by tresf** [comment]
I don't find it clumsy at all.  I very rarely add a high-note.  What are you proposing?

**Comment by Spekular** [comment]
@tresf that the max allowed time between clicks be lowered, or double clicking only opens piano roll when it happens outside the tombstones. Either that, some other improvement, or this behavior is removed and/or replaced with something else. I have many times accidentally opened a pattern in piano roll when I too quickly try to remove an accidental hit/note. When I do try to open piano roll on purpose, I'm frustrated by the unnecessary hit/note that is placed. Perhaps double clicking shouldn't update the tombstone that is clicked, so if there was a note there it is replaced on the second click and if there was'tit is remove on the second click.

**Comment by tresf** [comment]
> @tresf that the max allowed time between clicks be lowered

That is a system setting, QT just inherits the double-click speed of your OS.

> Or double clicking only opens piano roll when it happens outside the tombstones.

What you may decide to do is to put a counter and thus a delay on the QWidget's setVisible() and the first click listener.

Alternately, you may choose to ignore the first click in the piano roll always, but that might piss some people off... :smiley:

**Comment by Sti2nd** [comment]
> It's easy to accidentally do by adding a hit/note

Agreed. Double click is consistent throughout the software for opening Piano Roll and one of the fastest and best mouse interactions to use, so I doubt another key shortcut is the way to go.

**Comment by Spekular** [comment]
If #1681 is "fixed" so that a pattern will  be either piano or bb, with a one way conversion from bb to piano, I vote we remove double clicking as a shortcut to open piano roll. It's a much too minor action for such a major change. I'd also vote we a way to manipulate panning to bb-tracks.

**Comment by tresf** [comment]
> I vote we remove double clicking as a shortcut to open piano roll.

I don't see the problem with double-click.

**Comment by Spekular** [comment]
@tresf I feel like a pattern in bb-editor should be either an instrument pattern, bb-pattern, sample pattern(?) or automation pattern. There isn't really a reason to change between them, and I feel it should be a very deliberate action. It's some kind of gut feeling, I dunno.

**Comment by Sti2nd** [comment]
It could be discussed if editing beat notes in the Piano Roll is supported or not.

Now to the issue, there is two fields to double click in a B&B pattern. In the theme of LMMS 0.4.15, there was more space on the top which you could double click, and that does not create new steps. Here I have used Shift + drag on the empty area to increase the height of the pattern so you can see the top field.
![image](https://cloud.githubusercontent.com/assets/6373430/5881739/760e1990-a343-11e4-8ce9-632f578e8cfc.png)

I think a solution if possible is to make it so double clicking a step will not open the Piano Roll, only toggle it on and off, so to open Piano Roll you have to click at the narrow field at the top or use the context menu.

**Comment by Sti2nd** [comment]
I think it makes sense from a user perspective to differ from instrument patterns and drum patterns (although they are the same). Then one should perhaps open another Editor to edit drum patterns? A piano Roll without the scale and chord dropdown menu, and possibly without Q and automation too.

**Comment by tresf** [comment]
Step sequencers need the ability to be upgraded to Piano Rolls.  It's simply a natural progression in sequencing a track.  Fruityloops allows this too.  They have a slightly different approach as they allow the switching of keys without the changing of lengths via this drop-down hybrid approach:

![image](https://cloud.githubusercontent.com/assets/6345473/5882882/90d3498a-a31b-11e4-9832-1ac854b655db.png)


If double-clicking isn't wanted, you need to explain that, but the upgrade to a Piano Roll (whether through a context menu or double clicking) won't be removed.  It is simply too useful.

**Comment by tresf** [comment]
> A Piano Roll without the scale and chord dropdown menu, and possibly without Q and automation too.

Yeah, what you are describing is what FLStudio does (per screenshot above).

**Comment by Sti2nd** [comment]
https://github.com/LMMS/lmms/issues/1682#issuecomment-71258171
I was thinking that double clicking the steps could open this Drum Roll (instead of Piano Roll, lol), and double clicking on the area described in my comment ^ makes you able to open Piano Roll. So if you want to upgrade to an instrument pattern you simply opens it in Piano Roll.

**Comment by Sti2nd** [comment]
Should we perhaps make a new issue for this modified Piano Roll for drum beats?

**Comment by tresf** [comment]
> Should we perhaps make a new issue for this modified Piano Roll for drum beats?

I'd rather not until after #1681 is resolved.

**Comment by badosu** [comment]
@tresf @Sti2nd Did we consolidate what's this issue is about? It's not clear to me.

**Comment by tresf** [comment]
He wants double-click disabled for auto-opening the Piano Roll when using the step sequencer.  I personally use the context-menu so it wouldn't affect my usage, but I also don't think I have the double-click nuance happening as often on my setups... :confused:

**Comment by Sti2nd** [comment]
> It's not clear to me.

In B&B Editor, double click on a step, and Piano Roll will open. Also you have created a step, which weren't your intention, your intention was to open Piano Roll. That is the problem.

Possible solutions
- Differ between steps and patterns. The steps are represented by an icon from the theme, so beat patterns in the B&B Editor are regular Song Editor Patterns with the steps laid over. When you click the steps you can only toggle it on/off, when you click the pattern behind you can open Piano Roll.
![image](https://cloud.githubusercontent.com/assets/6373430/5937808/83ec0266-a6f4-11e4-98ea-ac2cf512b643.png)
- New key shortcut for opening a B&B pattern in Piano Roll.
- Remove double click functionality at all in B&B Editor. Not really a solution as you remove the whole feature...

**Comment by tresf** [comment]
@Sti2nd do you and @Spekular feel this double-click behavior is severe enough to disable?  If so, we should consider changing functionality.  If not, we should probably close this out as wont-fix.

**Comment by Sti2nd** [comment]
> severe enough to disable?

Not really. I can (perfectly fine) double click on the top area. Wouldn't Spekular simply using the context menu instead be a better solution than removing the whole feature to fix a "bug"?

> close this out as wont-fix.

But it is fixable? The first point in my last comment is possible, and the best approach with the current GUI. I realize, though, that it is somewhat strange to design the GUI to have an easy way to open Piano Roll. So I think we have a GUI error. Either the step icons should be smaller again so it would actually make sense to talk about double clicking on the upper field of the pattern, or they could just take up all the of the height in the pattern, and we could possibly implement a shortcut.

**Comment by tresf** [comment]
The upper field proposal won't happen.  It doesn't even make sense to me personally from a UI perspective.  

We either keep this

**Comment by Spekular** [comment]
@tres we at least ought to remove/restore the note that gets added/removed.

**Comment by Sti2nd** [comment]
> The upper field proposal won't happen. It doesn't even make sense to me personally from a UI perspective.

That whas the conclusion I reached too. I pointed out that if you remove this feauture one should perhaps increase the height of the step icons as the GUI seems to be designed around this feature.

**Comment by tresf** [comment]
> one should perhaps increase the height of the step icons as the GUI seems to be designed around this feature.

Sure, but that is a separate issue entirely. :wine_glass: 

> @tres we at least ought to remove/restore the note that gets added/removed.

Oh... You're talking about a single click operation being invoked during a double click action.  That's bug worthy.  Please re-title and adjust the bug report if you're good with that being the resolution.

-Tres

**Comment by tresf** [comment]
Nevermind, I edited it for you.  Sorry if this is what you meant to begin with. :beer:

**Comment by tresf** [comment]
Closed via #1783.  We decided to remove double-click.  It's the most sensible solution.  Sorry for those that really liked it. :confused:
